### DISCUSSION TITLE
LLHLS DVR (Live Rewind)

### DISCUSSION THREAD

**Discussion by getroot** [OvenMediaLabs/OvenMediaEngine#1074] [Announcement]
LLHLS DVR was added in today's 0.14.18 release. This allows viewers to rewind LLHLS live and replay past footage.

![image](https://user-images.githubusercontent.com/17779085/209091407-f8b0046a-c07a-429c-b40d-9927e80d9f88.png)

**_Please note that the ability to display the time of the seekable range in the OvenPlayer has not been updated yet. It will be updated soon. @SangwonOh_** 

To enable this function, add `<DVR>` to the settings as follows.
```
<Publishers>
	<LLHLS>
                ...
		<DVR>
			<Enable>false</Enable>
			<TempStoragePath>/tmp/ome_dvr/</TempStoragePath>
			<MaxDuration>3600</MaxDuration> <!-- Seconds -->
		</DVR>
	</LLHLS>
```

**TempStoragePath**
&emsp;A place where segments are temporarily stored to prevent excessive use of memory.

**MaxDuration**
&emsp;The duration to save segments, and the unit is seconds.

**Comment by naanlizard** [comment]
This is extremely exciting. Do we need to enable anything in Ovenplayer for this? I see we need to update it, so I will do that

Fantastic work, we will test this ASAP

**Comment by SangwonOh** [comment]
@naanlizard Hi! There is nothing to enable for this feature. As you said we only need new version of player.

**Comment by naanlizard** [comment]
Ok we have updated our server and are testing with the demo player. 

At first we forgot to switch DVR on, but using the new player still showed the DVR time bar. If you leave the page open, you appear to be able to seek around presumably using the cache in the browser. Is this a potential high memory usage/memory leak problem?

Otherwise so far it's smooth. We'll be testing with a 12 hour DVR to start

**Comment by SangwonOh** [comment]
Hi everyone. Displaying time in seekable range have been implemented on OvenPlayer `v0.10.25`.

![image](https://user-images.githubusercontent.com/21077319/210320265-821d4955-708a-4ece-b2c4-fc67369905dc.png)

**Comment by ChR-iSz** [comment]
Hi, nice job!

A short question: instead recording a stream -> it is possible to cut a part from dvr store, instead create a recording. 

So we can extract video from past dvr store, like as example "pastTimer"...

**Comment by Pinmonyoudom** [comment]
Hello @getroot, I would like to ask if OME team provides any REST API for LIVE DVR features, such as live rewind, seek backward/forward, or DVR playback control during livestreaming because now I can only config it via the `Server.xml` file.

**Reply by Pinmonyoudom** [reply]
Hello @getroot,

Just following up on my previous question. Does OME provide any REST API for LIVE DVR control (rewind, seek, DVR playback, etc.) during livestreaming, or is DVR configuration only available through `Server.xml`

Thank you.

**Reply by getroot** [reply]
This is mixing up two different things. DVR is a server-side feature you enable in Server.xml (<DVR> under the LLHLS/HLS publisher). It just makes OME retain old segments and list them in the .m3u8 playlist. Rewind/seek/playback is then handled entirely by the player, which seeks within that window by requesting older segments. There’s no REST API for playback control because in HLS/LLHLS that’s a client-side operation, not a server command.

**Reply by Pinmonyoudom** [reply]
Hello @getroot 

Oh my bad I want to know Does OME provide any REST API for LIVE DVR?

**Reply by qwerty13** [reply]
It's not by OME, it's by LLHLS protocol.
OME provided the hls DVR capabilities and then, you can seek to time by changing the offset of hls chunks you get.

For example on ffplay, you can use `-ss` parameter to play the stream with an offset:
`$ ffplay -ss {offset by seconds} "http://.../playlist.m3u8"`
(Note that you have to caculate the offset if you want to seek to an absolute datetime)

**Reply by Mengsreang-Chhoeung** [reply]
> This is mixing up two different things. DVR is a server-side feature you enable in Server.xml ( under the LLHLS/HLS publisher). It just makes OME retain old segments and list them in the .m3u8 playlist. Rewind/seek/playback is then handled entirely by the player, which seeks within that window by requesting older segments. There’s no REST API for playback control because in HLS/LLHLS that’s a client-side operation, not a server command.

To add to this from a server management perspective, I think having a REST API to dynamically configure or toggle the server-side DVR settings would be a fantastic feature for OME.

While playback control (seeking/rewinding) is absolutely a client-side operation handled via HLS/LLHLS as @getroot mentioned, being able to enable/disable DVR or adjust the retention window programmatically on the server—without needing a hard restart or manual edits to `Server.xml`—would be incredibly useful.

Is exposing the server-side DVR configuration to the OME REST API something the team might consider down the road?

**Reply by qwerty13** [reply]
> 
> To add to this from a server management perspective, I think having a REST API to dynamically configure or toggle the server-side DVR settings would be a fantastic feature for OME.
> 
> While playback control (seeking/rewinding) is absolutely a client-side operation handled via HLS/LLHLS as @getroot mentioned, being able to enable/disable DVR or adjust the retention window programmatically on the server—without needing a hard restart or manual edits to `Server.xml`—would be incredibly useful.
> 
> Is exposing the server-side DVR configuration to the OME REST API something the team might consider down the road?

I haven't tested it yet, but as it configured in `<Server><VirtualHosts><VirtualHost><Applications><Application><Publishers>` in the `Server.xml`, it must be configurable at publishers object of Application creation rest api body like others, not?

**Reply by Mengsreang-Chhoeung** [reply]
As I have tested, for the `<Server><VirtualHosts><VirtualHost><Applications><Application><Publishers>` in the `Server.xml` it's working properly for DVR, but for the REST API it doesn't work. Here is some of my testing:
**Endpoint:** POST http://localhost:8086/v1/vhosts
**Body:**
```json
[
    {
        "name": "39559267",
        "distribution": "39559267",
        "host": {
            "names": [
                "*"
            ]
        },
        "applications": [
            {
                "name": "room-1",
                "type": "live",
                "providers": {
                    "rtmp": {},
                    "srt": {},
                    "webrtc": {}
                },
                "publishers": {
                    "hls": {
                        "segmentDuration": 6,
                        "segmentCount": 5,
                        "dvr": {
                            "enable": true,
                            "tempStoragePath": "/data/hls",
                            "maxDuration": 36000
                        }
                    },
                    "llhls": {}
                },
                "outputProfiles": {
                    "outputProfile": [
                        {
                            "encodes": {
                                "audios": [
                                    {
                                        "bitrate": "128000",
                                        "channel": 2,
                                        "codec": "aac",
                                        "name": "aac_audio",
                                        "samplerate": 44100
                                    }
                                ],
                                "videos": [
                                    {
                                        "bitrate": "2000000",
                                        "bypassIfMatch": {
                                            "codec": "eq",
                                            "height": "lte",
                                            "width": "lte"
                                        },
                                        "codec": "h264",
                                        "framerate": 30,
                                        "height": 480,
                                        "name": "480p",
                                        "preset": "faster",
                                        "width": 854
                                    },
                                    {
                                        "bitrate": "4500000",
                                        "bypassIfMatch": {
                                            "codec": "eq",
                                            "height": "lte",
                                            "width": "lte"
                                        },
                                        "codec": "h264",
                                        "framerate": 60,
                                        "height": 720,
                                        "name": "720p60",
                                        "preset": "faster",
                                        "width": 1280
                                    }
                                ]
                            },
                            "name": "abr_stream",
                            "outputStreamName": "${OriginStreamName}",
                            "playlists": [
                                {
                                    "fileName": "master",
                                    "name": "master",
                                    "options": {
                                        "enableTsPackaging": true,
                                        "webRtcAutoAbr": false
                                    },
                                    "renditions": [
                                        {
                                            "audio": "aac_audio",
                                            "name": "480p",
                                            "video": "480p"
                                        },
                                        {
                                            "audio": "aac_audio",
                                            "name": "720p60",
                                            "video": "720p60"
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                }
            }
        ]
    }
]
```

**Result:**
```json
[
    {
        "message": "[Config] Unknown item found: application.publishers.hls.dvr",
        "statusCode": 400
    }
]
```

**Reply by qwerty13** [reply]
> ```json
>         "message": "[Config] Unknown item found: application.publishers.hls.dvr",
> ```

Ah so bad! both hls and llhls make this error...
